modules/qtwidgets/properties/transferfunctioneditor.cpp

```diff
@@ -453,69 +392,231 @@ void TransferFunctionEditor::keyPressEvent(QKeyEvent* keyEvent) {
+                    }
+                }
+                for (auto& item : groups_[group]) {
+                    item->setSelected(true);
+                }
+            });
+        }
+
+        // assign current selection to a group
+        for (auto&& i : util::make_sequence(1, 11, 1)) {
+            QString str = (i < 10 ? QString("Group &%1").arg(i) : "Group 1&0");
+            auto action = groupAssignMenu->addAction(str);
+            connect(action, &QAction::triggered, [ this, group = i % 10 ]() {
+                groups_[group].clear();
+                for (auto& item : getSelectedPrimitiveItems()) {
+                    groups_[group].push_back(item);
+                }
+            });
+        }
+    }
+
+    menu.addSeparator();
+    auto tfMenu = menu.addMenu("&Transfer Function");
+    {
+        auto clearTF = tfMenu->addAction("&Clear");
+        auto resetTF = tfMenu->addAction("&Reset");
+        auto importTF = tfMenu->addAction("&Import TF...");
+        auto exportTF = tfMenu->addAction("&Export TF...");
+
+        connect(clearTF, &QAction::triggered, this, [this]() {
+            NetworkLock lock(tfProperty_);
+            transferFunction_->clear();
+        });
+        connect(resetTF, &QAction::triggered, this, [this]() {
+            NetworkLock lock(tfProperty_);
+            tfProperty_->resetToDefaultState();
+        });
+        connect(importTF, &QAction::triggered, this, &TransferFunctionEditor::importTF);
+        connect(exportTF, &QAction::triggered, this, &TransferFunctionEditor::exportTF);
+    }
+
+    if (menu.exec(e->screenPos())) {
+        e->accept();
     }
-    transferFunction_->addPoint(vec2(pos.x() / width(), pos.y() / height()));
 }
 
-void TransferFunctionEditor::addControlPoint(QPointF pos, vec4 color) {
+void TransferFunctionEditor::addControlPoint(const QPointF& pos) {
+    dvec2 p(glm::clamp(pos.x() / width(), 0.0, 1.0), glm::clamp(pos.y() / height(), 0.0, 1.0));
+
+    NetworkLock lock(tfProperty_);
+    transferFunction_->add(p);
+}
+
+void TransferFunctionEditor::addControlPoint(double pos, const vec4& color) {
     // add control point to transfer function
-    if (pos.x() < 0.0) {
-        pos.setX(0.0);
-    } else if (pos.x() > width()) {
-        pos.setX(width());
-    }
-    NetworkLock lock;
-    transferFunction_->addPoint(static_cast<float>(pos.x() / width()), color);
+    pos = glm::clamp(pos / width(), 0.0, 1.0);
+
+    NetworkLock lock(tfProperty_);
+    transferFunction_->add(TFPrimitiveData{pos, color});
 }
 
-void TransferFunctionEditor::removeControlPoint(TransferFunctionEditorControlPoint* controlPoint) {
-    if (transferFunction_->getNumPoints() > 1) {
-        NetworkLock lock;
-        transferFunction_->removePoint(controlPoint->getPrimitive());
+void TransferFunctionEditor::addControlPointPeak(const QPointF& pos) {
+    dvec2 p(glm::clamp(pos.x() / width(), 0.0, 1.0), glm::clamp(pos.y() / height(), 0.0, 1.0));
```

Comment:
Can't pos be outside [0 1] now that they might not be normalized?

Reply:
Not yet, so far the editor, or rather Inviwo, only supports relative coords.